### ISSUE TITLE
yolov5-lite or yolov5-nano?

### ISSUE THREAD

**Issue by lifailifai** [ultralytics/yolov5#4462]
Thank your for great works, Do you have any plans to launch yolov5-lite or yolov5-nano? This is very important for mobile phones.

**Comment by glenn-jocher** [comment]
@zhunge all YOLOv5 models are natively mobile-compatible, i.e. see our TikTok with YOLOv5m running at 30 FPS on a 2 year old iPhone using the [YOLOv5 app](https://apps.apple.com/us/app/idetection/id1452689527).

https://vm.tiktok.com/ZMRL9Beo3/



<img width="400" src="https://user-images.githubusercontent.com/26833433/129928220-b48bb1ac-066d-4747-87c6-eae622d0d45e.png">

**Comment by litenjacob** [comment]
@glenn-jocher as someone who wants to run these models in the browser on Desktops and Phones, I think even lighter models (in size _and_ flops) can definitely be worth the tradeoffs. I'm currently testing out https://github.com/ppogg/YOLOv5-Lite 🤞

**Comment by glenn-jocher** [comment]
@litenjacob thanks for the feedback! The most basic type of model reduction on the smaller side would depend on the priority:

- smaller size: most easily attained by reduction/elimination of large output layers (i.e. P5-P6) or channel count/ops on large output layers, or depthwise convolutions on large output layers.
- faster speed: most easily attained by reduction/elimination of small output layers (i.e. P1-P3) or channel count/ops on small output layers, or strided convolutions on small output layers.

Note that smaller size and faster speed are not necessarily highly correlated, instead speed is dominated by small object layers while size is dominated by large object layers.

**Comment by litenjacob** [comment]
Ooh, thanks for elaborating – I hadn't realized that nuance before 🙏

I'm trying `yolov5` with `depth_multiple: 0.16` and `width_multiple: 0.37` now, which makes it fit in a single (< 4Mb) `tensorflowjs` shard with `int8` quantization. This loads fast enough and yields ~10 fps on my phone, which is good enough for my use-case.

I have noticed that it gives different/weird/broken results on iOS Safari compared to Python, Chrome, Safari Desktop and Firefox, but that feels like a bug in `tensorflowjs` for iOS, so I'll try to make a reduced test-case and report it there.

**Comment by ppogg** [comment]
> @glenn-jocher as someone who wants to run these models in the browser on Desktops and Phones, I think even lighter models (in size _and_ flops) can definitely be worth the tradeoffs. I'm currently testing out https://github.com/ppogg/YOLOv5-Lite 🤞

In fact, v5lite-s has not yet been fused in the conv+bn layer. I will upload the optimized code and model after writing the graduation paper, and it can be accelerated by 10%-15%

**Comment by glenn-jocher** [comment]
@ppogg actually fusing doesn't produce significant speedups in our experience. The graphs do become significantly shorter though, better for viewing with netron etc. Perhaps on other backends the story is different, but with PyTorch and CoreML we couldn't really observe any significant speedup. I'm sure it was there, it was probably just lost as a rounding error/measurement noise.

EDIT: still worthwhile to fuse before inference I think.

**Comment by ppogg** [comment]
@glenn-jocher  I am glad to receive your reply! What you said is correct.
As you said, It is so fast to use GPU inference. Only need 14-15ms per sample（2080ti，batch=1）,so fusing doesn't produce significant speedups in our experience on gpus. On some devices with poor computing power, like raspberry pi (arm) , or core i5 (x86), even core i3, the acceleration will be noticeable. Below is the test conducted on Core i5-10210U (1280*1280, 1000 sample)  :
`
Done. Average Sample:(0.629s). Total Sample:(1000)
`
`
Done. Average Sample:(0.557s). Total Sample:(1000)
`
The former is without  fusing, the latter is fusing.

I am not sure whether my comparison is correct, but I think the latency is higher on the Raspberry Pi, the contrast is more obvious.
My conclusion is consistent with you: Fusion is necessary. Attach one, the advantage is more obvious in the case of high latency

**Comment by glenn-jocher** [comment]
@ppogg good work on the profiling. Yes it seems like it's more important in low resource environments like raspberry pi. The speedup is noticeable in your CPU test also, nice!
